src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -759,3 +759,248 @@ - (void)invokeCommandWithEndpointID:(NSNumber *)endpointID
+            switch (eventTimeType) {
+            case MTREventTimeTypeSystemUpTime: {
+                if (!MTRInputIsStructurallyValid(itemDictionary, @ { MTREventSystemUpTimeKey : NSNumber.class })) {
+                    MTR_LOG_ERROR("Event report claims system uptime timing but does not have the time: %@", itemDictionary);
+                    return NO;
+                }
+                break;
+            }
+            case MTREventTimeTypeTimestampDate: {
+                if (!MTRInputIsStructurallyValid(itemDictionary, @ { MTREventTimestampDateKey : NSDate.class })) {
+                    MTR_LOG_ERROR("Event report claims epoch timing but does not have the time: %@", itemDictionary);
+                    return NO;
+                }
+                break;
+            }
+            default:
+                MTR_LOG_ERROR("Uknown time type for event report: %@", itemDictionary);
+                return NO;
+            }
+        }
+    }
+
+    return YES;
+}
+
+static BOOL MTRInvokeResponseIsStructurallyValid(id input)
+{
+    // Input is an array with a single value that must have MTRCommandPathKey.
+    if (!MTRInputIsStructurallyValid(input, @[ @ { MTRCommandPathKey : MTRCommandPath.class } ])) {
+        MTR_LOG_ERROR("Invoke response is not an array with the right things in it: %@", input);
+        return NO;
+    }
+
+    NSArray * inputArray = input;
+    if (inputArray.count != 1) {
+        MTR_LOG_ERROR("Invoke response is not an array with exactly one entry: %@", input);
+        return NO;
+    }
+
+    // Note that we have already checked the one entry is a dictionary that has
+    // MTRCommandPathKey.
+    NSDictionary * responseValue = inputArray[0];
+    id data = responseValue[MTRDataKey];
+    id error = responseValue[MTRErrorKey];
+
+    if (data != nil && error != nil) {
+        MTR_LOG_ERROR("Invoke response claims to have both data and error: %@", responseValue);
+        return NO;
+    }
+
+    if (error != nil) {
+        return MTRInputIsStructurallyValid(error, NSError.class);
+    }
+
+    if (data == nil) {
+        // This is valid: indicates a success status response.
+        return YES;
+    }
+
+    if (!MTRInputIsStructurallyValid(data, MTRInputStructureDataValue)) {
+        MTR_LOG_ERROR("Invoke response claims to have data that is not a data-value: %@", data);
+        return NO;
+    }
+
+    // Now we know data is a dictionary (in fact a data-value).  The only thing
+    // we promise about it is that it has type MTRStructureValueType.
+    NSDictionary * dataDictionary = data;
+    if (dataDictionary[MTRTypeKey] != MTRStructureValueType) {
+        MTR_LOG_ERROR("Invoke response data is not of structure type: %@", data);
+        return NO;
+    }
+
+    return YES;
+}
+
+BOOL MTRInputIsStructurallyValid(id input, id structure)
+{
+    if ([structure isEqual:MTRInputStructureDataValue]) {
```

Comment:
I would compare these magic values using `==` and declare the constants as `id` rather than `NSString *` since they're meant to be opaque to clients of this API.

Reply:
Magic strings went away.